### PR TITLE
Make crypto.decryptMessage return decryption results

### PR DESCRIPTION
**Pull request by richvdh**
... instead of having it call event.setClearData.

The main advantage of this is that it fixes a race condition, wherein apps
could see `event.isDecrypting()` to be true, but in fact the event had been
decrypted (and there was no `Event.decrypted` event on its way).

We're also fixing another race, wherein if the first attempt to decrypt failed,
a call to `attemptDecryption` would race against the first call and a second
attempt to decrypt would never happen.

It also gives a cleaner interface to MatrixEvent, at the expense of making the
`megolm` unit test a bit more hoop-jumpy.

### CODE REVIEW

File: spec/unit/event.spec.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+            let callCount = 0;
+
+            const crypto = {
+                decryptEvent: function() {
+                    ++callCount;
+                    console.log(`decrypt: ${callCount}`);
+                    if (callCount == 1) {
+                        // schedule a second decryption attempt while
+                        // the first one is still running.
+                        encryptedEvent.attemptDecryption(crypto);
```
Comment (Kegsay) [inline spec/unit/event.spec.js]: Can you please also verify that when this `Promise` resolves it has the cleartext.

File: src/crypto/algorithms/megolm.js [hunk 2]
```diff
@@ -535,18 +535,12 @@ utils.inherits(MegolmDecryption, base.DecryptionAlgorithm);
...
  */
-MegolmDecryption.prototype.decryptEvent = function(event) {
-    return this._decryptEvent(event, true);
-};
-
-
-// helper for the real decryptEvent and for _retryDecryption. If
-// requestKeysOnFail is true, we'll send an m.room_key_request when we fail
-// to decrypt the event due to missing megolm keys.
-MegolmDecryption.prototype._decryptEvent = async function(event, requestKeysOnFail) {
```
Comment (Kegsay) [inline src/crypto/algorithms/megolm.js]: Why are we removing the `requestKeysOnFail` flag?

File: src/crypto/index.js [hunk 3]
```diff
@@ -1295,6 +1296,25 @@ class IncomingRoomKeyRequestCancellation {
...
 
 /**
+ * The result of a (successful) call to decryptEvent.
+ *
+ * @typedef {Object} EventDecryptionResult
+ *
+ * @property {Object} clearEvent The plaintext payload for the event
+ *     (typically containing <tt>type</tt> and <tt>content</tt> fields).
+ *
+ * @property {string?} senderCurve25519Key Key owned by the sender of this event.
```
Comment (Kegsay) [inline src/crypto/index.js]: You need to start with a `?` not end with one. http://usejsdoc.org/tags-type.html

**Comment by Kegsay** [comment]
Otherwise LGTM, shame I failed to spot this race before :/

File: src/crypto/algorithms/megolm.js (hunk 2 above)
Comment (richvdh) [inline src/crypto/algorithms/megolm.js]: https://github.com/matrix-org/matrix-js-sdk/pull/523/commits/9550bca099b32059f54314f68f887f51dda3c0a9

File: spec/unit/event.spec.js (hunk 1 above)
Comment (richvdh) [inline spec/unit/event.spec.js]: well, it won't, because we return straight away at https://github.com/matrix-org/matrix-js-sdk/pull/523/files#diff-c2b46f9b1e3be0b9100ad9ef4970a4e3R358. I'm not keen to have to maintain code to keep track of a promise which we can return there when nothing relies on the result of `attemptDecryption` except the tests.

File: src/crypto/index.js (hunk 3 above)
Comment (richvdh) [inline src/crypto/index.js]: thanks. fixed.

File: spec/unit/event.spec.js (hunk 1 above)
Comment (Kegsay) [inline spec/unit/event.spec.js]: Having a function which returns different promises like this is incredibly error prone, regardless of its use solely in tests or not. If you don't want to fix the API then at the very least you should be documenting this edge case.

Comment (richvdh) [inline spec/unit/event.spec.js]: So I did this "properly" in 38c9a05.

**Comment by richvdh** [comment]
Hope this is close enough to the thing kegan LGTMed to be landed.
